docs/developer-guide/partition-data.rst

```diff
@@ -10,26 +10,76 @@ of the join have the same configuration settings for partitions.
 Co-partitioning Requirements
 ****************************
 
-* The input records for the join must have the :ref:`same keying scheme <keys-have-same-keying-scheme>`
+* The input records for the join must have the :ref:`same key schema <keys-have-same-key-schema>`
 * The input records must have the :ref:`same number of partitions <keys-have-same-number-of-partitions>` on both sides.
 * Both sides of the join must have the :ref:`same partitioning strategy <records-have-same-partitioning-strategy>`.  
 
 When your inputs are co-partitioned, records with the same key, from both
 sides of the join, are delivered to the same stream task during processing.
-If your inputs aren't co-partitioned, you need to :ref:`re-key one of the
-them <ensure-co-partitioning>` by using the PARTITION BY clause.
 
-.. _keys-have-same-keying-scheme:
+.. _keys-have-same-key-schema:
 
-Records Have the Same Keying Scheme
-===================================
+Records have the Same Key Schema
+================================
 
-The input records for the join must have the same keying scheme, which means
-that the join must use the same key field on both sides.
+For a join to work the keys from both sides must have the same serialized binary data.
 
-For example, you can join a stream of user clicks that's keyed by a ``VARCHAR userId``
-field with a table of user profiles that's keyed by a ``VARCHAR userId`` field. 
-The join won't match if the key fields don't have the same name and type.
+For example, you can join a stream of user clicks that's keyed on a ``VARCHAR`` user id with a table

```

Comment:
Couple of things here:
1. You can technically have key fields that are non-`VARCHAR` columns, and join on them, as long as the data in the kafka record key is the data in the key column expressed as a UTF-8 encoded string. I'm fine with us not calling this out as explicitly supported if we think it makes things confusing.
2. I think that it's a bit misleading to say that as long as the keys are both `VARCHAR` you can join. It really depends on the underlying serialization format - you wouldn't be able to join a `VARCHAR` column encoded as a JSON with one encoded using AVRO. We should specify that the keys are expected to be utf-8 encoded strings.

Reply:
1. Yeah, I know - deliberately didn't call it out as a) it is confusing, and b) I'm trying to get ride of this functionality.
2. Good point - I'll call this out.

Reply:
To make sure I've understood correctly, you're saying it's not possible to join two streams/tables with different serialization formats? I had no idea -- good thing we're calling this out in the docs!

Is this limitation from Kafka Streams, or specific to KSQL?

Was it the case that prior to the work on primitive keys, KSQL assumed/required keys to be strings serialized using Kafka's built-in string serializer? (Trying to understand why I haven't hit across this before.)

Reply:
> To make sure I've understood correctly, you're saying it's not possible to join two streams/tables with different serialization formats? I had no idea -- good thing we're calling this out in the docs!

It's just that the keys need to be in the same format

> Was it the case that prior to the work on primitive keys, KSQL assumed/required keys to be strings serialized using Kafka's built-in string serializer? (Trying to understand why I haven't hit across this before.)

Yes